File: Civitas-Ruby/lib/vista_textual.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

#encoding:utf-8

require 'io/console'

require_relative 'operaciones_juego'
require_relative 'civitas_juego'
require_relative 'respuestas'
require_relative 'operacion_inmobiliaria'
require_relative 'diario'
require_relative 'salidas_carcel'

module Civitas
  class Vista_textual
    attr_reader :iGestion, :iPropiedad
    
    def initialize
      @iPropiedad = -1
      @iGestion = -1
    end
    
    def mostrar_estado(estado)
      puts estado
    end
    
    def pausa
      print "Pulsa una tecla"
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end
    
    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l
        index += 1
      }

      opcion = lee_entero(lista.length, tab+'Elige una opcion: ', tab+'Valor erroneo')
      return opcion
    end
    
    def comprar
      opciones = Array.new
      opciones = ['Si', 'No']
      opcion = menu('Desea comprar la calle', opciones)
      
      lista_respuestas = [Civitas::Respuestas::SI, Civitas::Respuestas::NO]
      
      lista_respuestas[opcion]
    end

    def gestionar      
      propiedades = Array.new
      
      for t in @juegoModel.getjugadoractual.propiedades
        propiedades << t.nombre
      end
            
      opciones = Array.new
      opciones = ['Vender', 'Hipotecar', 'Cancelar Hipoteca', 'Construir Casa',
        'Construir Hotel', 'Terminar']
      @iGestion = menu('Que gestion inmobiliaria desea hacer', opciones)
      
      if !propiedades.empty?
        @iPropiedad = menu("Que propiedad desea gestionar", propiedades)
      else
        puts 'No tienes propiedades'
      end
      
    end

    def get_gestion
      @iGestion
    end

    def get_propiedad
      @iPropiedad
    end

    def mostrar_siguiente_operacion(operacion)
      puts operacion.to_s
    end

    def mostrar_eventos
      while Diario.instance.eventos_pendientes
        puts Diario.instance.leer_evento
      end
    end
    
    def salir_carcel
      opciones = Array.new
      opciones = ['Pagando', 'Tirando el dado']
      opcion = menu('Elige la forma para intentar salir de la carcel', 
               opciones)
             
      lista_salidas = [Civitas::SalidasCarcel::PAGANDO, Civitas::SalidasCarcel::TIRANDO]
      
      lista_salidas[opcion]
    end

    def set_civitas_juego(civitas)
        @juegoModel=civitas
        self.actualizar_vista
    end

    def actualizar_vista
      jugador_actual = @juegoModel.getjugadoractual
      casilla = @juegoModel.getcasillaactual      
      
      #puts jugador_actual.to_s
      #puts casilla.to_s  No muestra la casilla    
    end    
  end
end

File: Civitas-Java/Civitas/src/main/java/juegoTexto/VistaTextual.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package juegoTexto;

import civitas.CivitasJuego;
import civitas.Diario;
import civitas.OperacionesJuego;
import civitas.SalidasCarcel;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import civitas.Casilla;
import civitas.Jugador;
import civitas.TituloPropiedad;

/**
 * @author elena
 * @date 25/10/2019
 */
class VistaTextual {
  
    CivitasJuego juegoModel; 
    int iGestion=-1;
    int iPropiedad=-1;
    private static String separador = "=====================";
  
    private Scanner in;
  
    VistaTextual(){
        in = new Scanner (System.in);
    }
  
    void mostrarEstado(String estado){
        System.out.println (estado);
    }
              
    void pausa(){
        System.out.print ("Pulsa una tecla");
        in.nextLine();
    }

    int leeEntero(int max, String msg1, String msg2) {
        boolean ok;
        String cadena;
        int numero = -1;
        do{
            System.out.print (msg1);
            cadena = in.nextLine();
        try{  
            numero = Integer.parseInt(cadena);
            ok = true;
        } catch(NumberFormatException e){ // No se ha introducido un entero
            System.out.println (msg2);
            ok = false;  
        }
        if(ok && (numero < 0 || numero >= max)){
            System.out.println (msg2);
            ok = false;
        }
        } while (!ok);

    return numero;
  }

    int menu(String titulo, ArrayList<String> lista) {
        String tab = "  ";
        int opcion;
        System.out.println(titulo);
        for (int i = 0; i < lista.size(); i++) {
            System.out.println(tab+i+"-"+lista.get(i));
        }

        opcion = leeEntero(lista.size(), "\n"+tab+"Elige una opción: ",
                           tab+"Valor erróneo");
        return opcion;
    }

    SalidasCarcel salirCarcel() {
        int opcion = menu ("Elige la forma para intentar salir de la carcel",
        new ArrayList<> (Arrays.asList("Pagando","Tirando el dado")));
        return (SalidasCarcel.values()[opcion]);
    }

    Respuestas comprar(){
        int opcion = menu("¿Desea comprar esta calle?", 
                    new ArrayList<> (Arrays.asList("Si","No")));
        return(Respuestas.values()[opcion]);
    }

    void gestionar(){
        ArrayList<String> propiedades = new ArrayList();
        
        //He modificado la visibilidad del método getPropiedades y getNombre
        for(TituloPropiedad t : juegoModel.getJugadorActual().getPropiedades()){
            propiedades.add(t.getNombre());
        }
        
        iGestion = menu("¿Qué gestión inmobiliaria desea hacer?", 
                        new ArrayList<> (Arrays.asList("Vender","Hipotecar",
                        "Cancelar Hipoteca", "Construir Casa", "Construir hotel",
                        "Terminar")));
        
        if(iGestion != 5){
            iPropiedad = menu("¿Qué propiedad desea gestionar?", propiedades);
        }
                
    }

    public int getGestion(){ return iGestion; }
    public int getPropiedad(){ return iPropiedad; }

    void mostrarSiguienteOperacion(OperacionesJuego operacion){
        System.out.println(operacion);
    }

    void mostrarEventos(){
        while(Diario.getInstance().eventosPendientes()){
            System.out.println(Diario.getInstance().leerEvento());
        }
    }
  
    public void setCivitasJuego(CivitasJuego civitas){ 
        juegoModel=civitas;
        this.actualizarVista();
    }
  
    void actualizarVista(){
        Casilla casilla = juegoModel.getCasillaActual();
        Jugador jugadorActual = juegoModel.getJugadorActual();
        System.out.println("\n" + casilla.toString() + "\n" + jugadorActual.toString());
    } 
}

